File: src/main/java/com/ririv/quickoutline/textProcess/methods/parser.java
package com.ririv.quickoutline.textProcess.methods;

import com.ririv.quickoutline.model.Bookmark;

import java.util.ArrayList;
import java.util.List;

public interface parser {

    // linearBookmarkList 传递是为了可以使该行根据上文判断该行level
    Bookmark parseLine(int offset, String line, List<Bookmark> linearBookmarkList);

    default List<Bookmark> parse(List<String> text, int offset){
        List<Bookmark> linearBookmarkList = new ArrayList<>();
        for (String line : text) {
            var current = parseLine(offset, line, linearBookmarkList);
            linearBookmarkList.add(current);
        }
        return linearBookmarkList;
    }
}

File: src/lib/outlineParser/Parser.ts
// Ported from parser.java
import type { BookmarkUI } from "../../components/bookmark/types";

export interface Parser {
    /**
     * Parses a single line of text into a Bookmark object.
     * @param line The text line to parse.
     * @param linearBookmarkList The list of bookmarks parsed so far (for context).
     */
    parseLine(line: string, linearBookmarkList: BookmarkUI[]): BookmarkUI;

    /**
     * Parses a list of text lines into a list of Bookmark objects.
     * @param text The list of text lines.
     */
    parse(text: string[]): BookmarkUI[];
}

/**
 * Abstract base class to provide the default `parse` implementation.
 */
export abstract class BaseParser implements Parser {
    abstract parseLine(line: string, linearBookmarkList: BookmarkUI[]): BookmarkUI;

    parse(text: string[]): BookmarkUI[] {
        const linearBookmarkList: BookmarkUI[] = [];
        for (const line of text) {
            try {
                const current = this.parseLine(line, linearBookmarkList);
                linearBookmarkList.push(current);
            } catch (e) {
                // In Java, BookmarkFormatException is thrown.
                // We propagate the error so the UI can handle it or the parser stops.
                throw e;
            }
        }
        return linearBookmarkList;
    }
}
